eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala

```diff
@@ -74,298 +48,319 @@ object Sphinx extends Logging {
+      * @param payloads      payloads for all the hops.
+      * @return filler bytes.
+      */
+    def generateFiller(keyType: String, sharedSecrets: Seq[ByteVector32], payloads: Seq[ByteVector]): ByteVector = {
+      require(sharedSecrets.length == payloads.length, "the number of secrets should equal the number of payloads")
+
+      (sharedSecrets zip payloads).foldLeft(ByteVector.empty)((padding, secretAndPayload) => {
+        val (secret, perHopPayload) = secretAndPayload
+        val perHopPayloadLength = peekPayloadLength(perHopPayload)
+        require(perHopPayloadLength == perHopPayload.length + MacLength, s"invalid payload: length isn't correctly encoded: $perHopPayload")
+        val key = generateKey(keyType, secret)
+        val padding1 = padding ++ ByteVector.fill(perHopPayloadLength)(0)
+        val stream = generateStream(key, PayloadLength + perHopPayloadLength).takeRight(padding1.length)
+        padding1.xor(stream)
+      })
+    }
 
-  /**
-    * Builds an encrypted onion packet that contains payloads and routing information for all nodes in the list
-    *
-    * @param sessionKey     session key
-    * @param publicKeys     node public keys (one per node)
-    * @param payloads       payloads (one per node)
-    * @param associatedData associated data
-    * @return an OnionPacket(onion packet, shared secrets). the onion packet can be sent to the first node in the list, and the
-    *         shared secrets (one per node) can be used to parse returned error messages if needed
-    */
-  def makePacket(sessionKey: PrivateKey, publicKeys: Seq[PublicKey], payloads: Seq[ByteVector], associatedData: ByteVector32): PacketAndSecrets = {
-    val (ephemeralPublicKeys, sharedsecrets) = computeEphemeralPublicKeysAndSharedSecrets(sessionKey, publicKeys)
-    val filler = generateFiller("rho", sharedsecrets.dropRight(1), PayloadLength + MacLength, MaxHops)
+    /**
+      * Decrypt the incoming packet, extract the per-hop payload and build the packet for the next node.
+      *
+      * @param privateKey     this node's private key.
+      * @param associatedData associated data.
+      * @param packet         packet received by this node.
+      * @return a DecryptedPacket(payload, packet, shared secret) object where:
+      *         - payload is the per-hop payload for this node.
+      *         - packet is the next packet, to be forwarded using the info that is given in the payload.
+      *         - shared secret is the secret we share with the node that sent the packet. We need it to propagate
+      *         failure messages upstream.
+      *         or a BadOnion error containing the hash of the invalid onion.
+      */
+    def peel(privateKey: PrivateKey, associatedData: ByteVector, packet: wire.OnionRoutingPacket): Either[wire.BadOnion, DecryptedPacket] = packet.version match {
+      case 0 => Try(PublicKey(packet.publicKey, checkValid = true)) match {
```

Comment:
The field `publicKey` isn't typed in `OnionRoutingPacket` because we want to handle the "invalid pubkey" case at this layer.

But for optimization reasons (it is pretty expensive to fully validate a public key), our codec currently doesn't do any validation: we just read 33 bytes and ensure that the key begins with `02` or `03`.

How about we properly type the field to `PublicKey` in `OnionRoutingPacket`, and we do a proper validation here (that is, we verify that the point is indeed on the curve here).

The tradeoff here would be that un unparseable key would probably not even start with a `02` or `03`, and the failure would happen in the codec. Okay, not sure about this.

@sstone WDYT?

Reply:
We definitely need to verify here that the point is on the curve (which is done with the `checkValid = true` flag. So if we do validation here, it feels a bit weird to do partial validation in the codec...
Especially since that means that we need to make the codec more complex to correctly return the right `BadOnion` error in case the public key is invalid (not sure it's easy to do with the current decoding flow).

Reply:
It bothers me a bit too to not use types here but it makes parsing and error handling easy to read and compare to the BOLTs so I can live with it. 
Also our process is already split in 2 and we do not check the htlc's onion packet until we've received a revocation so using a stricter type should probably be part of a larger refactoring where the htlc is fully parsed when we receive it